pallets/author/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019-2020 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Block author tracking by inherents
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{decl_error, decl_module, decl_storage, ensure, weights::Weight};
+use frame_system::{ensure_none, Config as System};
+use parity_scale_codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use sp_inherents::ProvideInherentData;
+use sp_inherents::{InherentData, InherentIdentifier, IsFatalError, ProvideInherent};
+use sp_runtime::RuntimeString;
+use sp_std::vec::Vec;
+
+pub trait EventHandler<Author> {
+	/// Note that the given account ID is the author of the current block.
+	fn note_author(author: Author);
+}
+
+pub trait IsValidator<AccountId> {
+	fn is_validator(account: AccountId) -> bool;
+}
+
+pub trait Config: System {
+	/// An event handler for authored blocks.
+	type EventHandler: EventHandler<Self::AccountId>;
+	/// Checks if account is a validator
+	type IsAuthority: IsValidator<Self::AccountId>;
+}
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// Author already set in block.
+		AuthorAlreadySet,
+		NotValidator,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as Author {
+		/// Author of current block.
+		Author: Option<T::AccountId>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		/// Inherent to set the author of a block
+		#[weight = 1_000_000]
+		fn set_author(origin, author: T::AccountId) {
+			ensure_none(origin)?;
+			ensure!(<Author<T>>::get().is_none(), Error::<T>::AuthorAlreadySet);
+			ensure!(T::IsAuthority::is_validator(author.clone()), Error::<T>::NotValidator);
```

Comment:
This check ensures that the author set in this pallet is always a member of the current validator set. This prevents incorrectly awarding points to non-validators in `stake` for block production (through `EventHandler` impl). It ensures that all points awarded for block production only go to current validators.

I considered making `T::EventHandler::note_author` fallible so points are only awarded if the author is in the validator set recognized in `stake`, but this would lead to situations in which a non-validator is set as the block author in this pallet (`author`) and no one is awarded points. I feel this outcome would be more confusing than the current implementation, which prevents setting the author unless the author is in the recognized validator set in `stake`.

Reply:
I agree